src/source/source_cache.js

```diff
@@ -241,14 +241,17 @@ class SourceCache extends Evented {
             tile.state = 'errored';
             if ((err: any).status !== 404) this._source.fire(new ErrorEvent(err, {tile}));
             // continue to try loading parent/children tiles if a tile doesn't exist (404)
-            else this.update(this.transform);
```

Comment:
Previous version of code wasn't resulting with loaded parents as usedForTerrain == true and updateForTerrain == false would leave early:

https://github.com/mapbox/mapbox-gl-js/blob/941ba0fdb8d1b945dcd7040c129bff559ed5a5e5/src/source/source_cache.js#L472-L475

Now, with the change, tile size is not correct (it has to be the same tile size as supplied in terrain.js):

https://github.com/mapbox/mapbox-gl-js/blob/941ba0fdb8d1b945dcd7040c129bff559ed5a5e5/src/terrain/terrain.js#L281-L283

Also, it is essential to reset lookup and call line 

https://github.com/mapbox/mapbox-gl-js/blob/941ba0fdb8d1b945dcd7040c129bff559ed5a5e5/src/terrain/terrain.js#L283

right after sourceCache.update.